src/valiant/sgrna_utils.py

```diff
@@ -0,0 +1,129 @@
+########## LICENCE ##########
+# VaLiAnT
+# Copyright (C) 2020, 2021, 2022 Genome Research Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#############################
+
+from typing import Dict, List, Optional, Set
+import pandas as pd
+from .constants import META_MUT_POSITION, META_REF, META_PAM_MUT_SGRNA_ID, ARRAY_SEPARATOR
+
+
+def get_codon_index_range(codon_start: int, codon_end: int) -> List[int]:
+    return (
+        [codon_start] if codon_end == codon_start else
+        list(range(codon_start, codon_end + 1))
+    )
+
+
+def get_sgrna_ids(frame: int, codon_to_sgrna_ids: Dict[int, str], mut_pos: int, mut_ref: Optional[str]) -> Set[str]:
+    """List the identifiers of the sgRNA's whose variants map to codons altered by the mutation"""
+
+    def _get_codon_index(x: int) -> int:
+        # Assuming the positions are relative indices
+        return get_codon_index(frame, 0, x)
+
+    codon_start: int = _get_codon_index(mut_pos)
+    ref_length: int = len(mut_ref) if mut_ref else 0
+
+    return (
+        {
+            codon_to_sgrna_ids[codon_index]
+            for codon_index in get_codon_index_range(codon_start, _get_codon_index(mut_pos + ref_length - 1))
+            if codon_index in codon_to_sgrna_ids
+        } if ref_length > 1 else
+        {codon_to_sgrna_ids[codon_start]} if codon_start in codon_to_sgrna_ids else
+        set()
+    )
```

Comment:
Bit difficult to read - perhaps better to take some logic out of return statement?